pkg/cnb/project_descriptor.go

```diff
@@ -105,6 +107,35 @@ func processFiles(appDir string, d build) error {
 	})
 }
 
+// Because CNB Project Descriptor v0.2 has two ways for defining environment variables.
+// see https://github.com/buildpacks/spec/blob/main/extensions/project-descriptor.md#iobuildpacksbuildenv-optional
+// This function calculates the final environment variables following this logic:
+//  if (new V2) AND NOT (old V2) returns (new V2)
+// 	if (old V2) AND NOT (new V2) returns (old V2)
+// 	if (new V2) AND (old V2) returns (new V2) + (new V2 that overrides old V2) + (old V2 that are not defined in new V2)
+func getEnvVariables(d descriptorV2) []envVariable {
+	env := d.IO.Buildpacks.Env
+	if env == nil {
+		env = d.IO.Buildpacks.BuildEnv.Env
+	} else {
+		if d.IO.Buildpacks.BuildEnv.Env != nil {
+			// both env variables definition were used
+			env = mergeEnvVariables(env, d.IO.Buildpacks.BuildEnv.Env)
+		}
+	}
+	return env
+}
+
+func mergeEnvVariables(old []envVariable, new []envVariable) []envVariable {
```

Comment:
pack appears to just use the newer field instead of merge: https://github.com/buildpacks/pack/pull/1479/files#diff-4216f3c5d147f7361545afe9fb76c1fe23790f314e3f39ae1eca8559d726bd09R52

maybe this can help simplify things

anyone have opinions? @tomkennedy513 @samj1912 @matthewmcnew @ncarlson @yaelharel @chenbh

Reply:
Ya, i think if the newer field is populated just use that

Reply:
Do you mean to keep it how it was before?

Reply:
I think it was backwards before:

pack:
```
env := versionedDescriptor.IO.Buildpacks.Build.Env
if env == nil {
    env = versionedDescriptor.IO.Buildpacks.Env.Build
}
```

before:

```
env := d.IO.Buildpacks.Env
if env == nil {
	env = d.IO.Buildpacks.BuildEnv.Env
}
```

Reply:
Yeap, I added the logic before following @samj1912 [comment](https://github.com/pivotal/kpack/pull/1027#discussion_r961150010)